### ISSUE TITLE
Cuda EP parallelization issues for batches

### ISSUE THREAD

**Issue by admayber** [microsoft/onnxruntime#11047]
**Describe the bug**
We are attempting to use gpu to accelerate inference for batches of hundreds of images by exploiting gpu parallelism using the C++ API, and are having difficulty getting the cuda EP to parallelize our data in any way. There are two separate but related issues / questions.

First, we are not seeing any performance scaling above a batch of size of 8 - 10, it seems that above that size there is a lot of sequential processing and latency increases. For example, the run time of a batch of 100 seems to be about 10x that of a batch size of 10. This was very surprising since we expected ORT to be able to leverage gpu parallelism well beyond this (we're definitely not saturating the gpu).

_Question 1:_ When we did profiling with NSight we found that all of the operations were being executed in a single stream. Is this expected for all models, or does it potentially indicate some optimization issue?

To try and circumvent this, we tried making concurrent requests to the gpu session by making async calls (see code example below). Again, latency-wise it seemed that these requests were being serialized into a single stream, and profiling with NSight confirmed this.

_Question 2:_ Since it seems that this kind of explicit concurrency is supposed to be supported (https://github.com/microsoft/onnxruntime/discussions/7876), what are the correct settings / mechanisms needed to leverage it? Ideally we would not create multiple separate sessions as we want to reuse the same model weights and reduce gpu memory overhead.

**Urgency**
Blocking some time-sensitive projects that have strict latency requirements for large batching. Due dates start in about 4 weeks.

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- ONNX Runtime installed from (source or binary): binary
- ONNX Runtime version: 1.10-gpu
- Python version: N/A
- Visual Studio version (if applicable): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: CUDA 11.4, cuDNN 8.2.4
- GPU model and memory: V100, 16GB

**To Reproduce**
The partial example below outlines how we've tried to do concurrency in gpu:

```cpp
    Ort::Env env(OrtLoggingLevel::ORT_LOGGING_LEVEL_ERROR, "eval");
    OrtCUDAProviderOptions cudaOptions;
    cudaOptions.gpu_mem_limit = 6L * (1024L * 1024L * 1024L);

    Ort::SessionOptions sessionOptions;
    sessionOptions.SetGraphOptimizationLevel(GraphOptimizationLevel::ORT_ENABLE_ALL);
    sessionOptions.AppendExecutionProvider_CUDA(cudaOptions);

    // <...do batch prep here and create session here...>

   // launch concurrent calls and get futures (abstracted a bit)
   auto async1 = std::async(std::launch::async, doInference, std::ref(session), std::cref(batch1));
   auto async2 = std::async(std::launch::async, doInference, std::ref(session), std::cref(batch2));

    // join threads and get results
    auto res1 = async1.get();
    auto res2 = async2.get();
```

**Additional context**
This is a VIT model architecture, please reach out privately if additional details would be helpful. Attaching two NSight report files, one that shows the single-batch performance and one that shows two batches being serialized into a single stream (see readme for details).
[nsys-reports.zip](https://github.com/microsoft/onnxruntime/files/8375463/nsys-reports.zip)

**Comment by hariharans29** [comment]
`When we did profiling with NSight we found that all of the operations were being executed in a single stream. Is this expected for all models, or does it potentially indicate some optimization issue?`

This is expected. The CUDA EP only supports the sequential executor which means all of the operations are executed sequentially on the same stream. Some other EPs like the CPU EP support the parallel executor, a concept if applied to the CUDA EP would mean that we could potentially use different streams to parallelize node executions but enabling this is non-trivial for the CUDA EP and is currently not in the pipeline.

`To try and circumvent this, we tried making concurrent requests to the gpu session by making async calls (see code example below). Again, latency-wise it seemed that these requests were being serialized into a single stream, and profiling with NSight confirmed this.`

Currently, there is only one stream per CUDA EP and unfortunately due to this the concurrent requests are being serviced by this single stream. We plan to enable per-thread streams in the CUDA EP which should mean that concurrent requests can be serviced by different streams in parallel.

**Comment by dashesy** [comment]
Perhaps it would be a little easier if different sessions (different processes) could use different streams? Then we could parallelize by loading the model to memory in 2 processes?

**Comment by admayber** [comment]
Agreed with @dashesy, I have not tried this yet but support would be good if it doesn't currently exist

**Comment by hariharans29** [comment]
Just to be clear - different sessions do use different streams. It is just that within one session, we are limited to using one compute stream and this is affecting the concurrency aspect.
